### PR TITLE
Fix row-types on Quint operator conversion

### LINKED ISSUE #2542: Row variables are lost when converting Quint operator types
Writing integration tests for Quint  record operators I,be found that we seem to be dropping or ignoring the real variable that should track open records when converting the types of operators that take records as parameters.

### PR DESCRIPTION
**Pull request by shonfeder**
Fixes #2542

Adds a scoped builder constructor for row-based records.

This PR has also accumulated a few changes looking forward to #2552:

- c4c7f65
- 050f5ea

I realized just after updating the integration tests for these changes that the PR was big enough to benefit from revue and merge.

I'll resume work on #2552 in a followup PR.
 

- [x] Tests added for any new code
- [x] Ran `make fmt-fix` (or had formatting run automatically on all files edited)
- [x] Documentation added for any new functionality
- [x] [Entries added to `./unreleased/`][changelog format] for any new functionality

[changelog format]: https://github.com/informalsystems/apalache/blob/main/CONTRIBUTING.md#how-to-record-a-change

### CODE REVIEW

**Comment by shonfeder** [comment]
There’s an integration test failing because of a missing fixture generation (I think).

File: tlair/src/main/scala/at/forsyte/apalache/tla/typecomp/unsafe/UnsafeFunBuilder.scala [hunk 1]
```diff
@@ -22,6 +22,58 @@ class UnsafeFunBuilder extends ProtoBuilder {
   private val strBuilder = new UnsafeLiteralAndNameBuilder
   private def mkTlaStr: String => TlaEx = strBuilder.str
 
+  private def formRecordFieldType(args: Seq[TlaEx]): SortedMap[String, TlaType1] = {
```
Comment (shonfeder) [inline tlair/src/main/scala/at/forsyte/apalache/tla/typecomp/unsafe/UnsafeFunBuilder.scala:25]: This was just moved so the logic could be reused.

File: tla-io/src/test/scala/at/forsyte/apalache/io/quint/TestQuintEx.scala [hunk 2]
```diff
@@ -495,6 +524,31 @@ class TestQuintEx extends AnyFunSuite {
...
 
+  test("`with` operator conversion preserves row-typing") {
+    val typ = QuintRecordT.ofFieldTypes("a", ("s", QuintIntT()), ("t", QuintIntT()))
+    val rec = Q.app("Rec", Q.s, Q._1, Q.t, Q._2)(typ)
+    val exp = Q.quint.exToTla(rec).get
+    val expectedTlaType = RecRowT1(RowT1(VarT1("a"), ("s", IntT1), ("t", IntT1)))
+
+    assert(Quint.typeToTlaType(typ) == expectedTlaType)
+    assert(exp.typeTag == Typed(expectedTlaType))
+  }
```
Comment (bugarela) [inline tla-io/src/test/scala/at/forsyte/apalache/io/quint/TestQuintEx.scala:535]: This seems off place, I think it's an old version of the test added above? It doesn't seem to be testing the `with` operator.

File: tla-io/src/test/scala/at/forsyte/apalache/io/quint/TestQuintEx.scala [hunk 3]
```diff
@@ -495,6 +524,31 @@ class TestQuintEx extends AnyFunSuite {
...
+
+  test("operator def conversion preserves row-typing") {
+    // def updateF1(r) : {f1: int | a} => {s: int | a} = r.with("f1", 1)
```
Comment (bugarela) [inline tla-io/src/test/scala/at/forsyte/apalache/io/quint/TestQuintEx.scala:538]: This signature is actually not type correct. I'm not sure if this is relevant for the test here. A simple way to make it correct is replacing `s` with `f1`, but perhaps you want something like this:
```
def updateF1(r) : { f1: int | a } => { s: int } = { s: r.f1 }
```

**Review by bugarela — APPROVED** [review]
LGTM! Just a couple of comments on the unit tests

File: tla-io/src/test/scala/at/forsyte/apalache/io/quint/TestQuintEx.scala (hunk 2 above)
Comment (shonfeder) [reply tla-io/src/test/scala/at/forsyte/apalache/io/quint/TestQuintEx.scala:535]: Oh yeah. That was an oversight, this got covered in the best below. Thanks!

File: tla-io/src/test/scala/at/forsyte/apalache/io/quint/TestQuintEx.scala (hunk 3 above)
Comment (shonfeder) [reply tla-io/src/test/scala/at/forsyte/apalache/io/quint/TestQuintEx.scala]: Typo there. Thanks!

**Pull request merged**